internal/controller/postgrescluster/pgbouncer.go

```diff
@@ -295,10 +288,34 @@ func (r *Reconciler) reconcilePGBouncerService(
 		TargetPort: intstr.FromString(naming.PortPGBouncer),
 	}}
 
+	err := errors.WithStack(r.setControllerReference(cluster, service))
+
+	return service, true, err
+}
+
+// +kubebuilder:rbac:groups="",resources="services",verbs={get}
+// +kubebuilder:rbac:groups="",resources="services",verbs={create,delete,patch}
+
+// reconcilePGBouncerService writes the Service that resolves to PgBouncer.
+func (r *Reconciler) reconcilePGBouncerService(
+	ctx context.Context, cluster *v1beta1.PostgresCluster,
+) (*corev1.Service, error) {
+	service, specified, err := r.generatePGBouncerService(cluster)
```

Comment:
❓ What made you want to have the generate function determine if we need the service? I would think reconciles purpose would be to check if a pgbouncer service is needed then use `generatePGBouncerService` get that service. 

Should generate care if pgbouncer service is asked for in the spec?

Reply:
With this split, generate does all the interpreting of the cluster spec and is the one place that decides the service name. Reconcile makes the API calls.

The `cluster.Spec.Proxy.PGBouncer == nil` guard is the crux of it for me. If we move that to reconcile, then it seems like generate should take a `PGBouncerPodSpec` (i.e. only call this when there's a spec). But generate also/still needs the cluster to set the metadata and ownership. Sending both seems confusing (does the bouncer spec have to match what's in the cluster spec?) or prone to mistakes (one line reads from the bouncer spec, another line reads from the cluster).